### DISCUSSION TITLE
UI freezing during long/streamed task execution

### DISCUSSION THREAD

**Discussion by LSXAxeller** [AvaloniaUI/Avalonia#15913] [Q&A] [answered]
The UI is freezing up when I try to send messages using `ChatService.Generate` in command `MessageSend()`. It happens when calling the method directly or with await, but it works fine with Task.Run. although this doesn't happen with other resource-intensive tasks. like loading large models to memory just using await with no need to Task.Run() or new Thread().

```
    [RelayCommand]
    private void MessageSend() // async Task MessageSend()
    {
        // Send the message to the chat session and add it to the chat session's messages
        var userInput = ChatUserInput;
        ChatUserInput = string.Empty;
    //    Task.Run(async () => await ChatService.Generate(userInput, GenerationParameters, PromptParameters) );
    //    new Thread((() => ChatService.Generate(userInput, GenerationParameters, PromptParameters))).Start();
         ChatService.Generate(userInput, GenerationParameters, PromptParameters); // await still freeze too
    }
```

**ChatService.Generate**
```
public async Task Generate(string userInput, TextGenerationParameters? generationParameters = null,
        TextPromptParameters? promptParameters = null)
    {
        if (_generation.State == GenerationState.Started)
        {
            await _generation.CancelToken?.CancelAsync()!;
            return;
        }

        if (string.IsNullOrEmpty(userInput))
        {
            logger.LogWarning("ChatUserInput is empty.");
            return;
        }

        if (_generation.SelectedGenerationModel == null)
        {
            logger.LogWarning("Text Generation Model isn't loaded yet.");
            return;
        }

        if (SelectedChatSession == null)
            await Create($"Chat {DateTime.Now}");


        _generation.State = GenerationState.Started;

        string prompt;
        _generation.CancelToken = new CancellationTokenSource();

        if (SelectedChatSession?.Executor == null ||
            SelectedChatSession.Executor.IsFirstRun(nameof(StatefulExecutorBase)).Result)
        {
            if (SelectedChatSession?.Executor == null)
                SelectedChatSession!.Executor = new InferenceService(
                    new InteractiveExecutor(_generation.Model!.CreateContext(_generation.ModelParameters!)));
            prompt =
                $"{(SelectedChatSession.Messages.Count > 0 ? SelectedChatSession.ToString() : string.Empty)} \nUser: {userInput}";
        }
        else
        {
            prompt = promptParameters?.Prompt(userInput, SelectedChatSession?.Messages.Count <= 2)!;
        }

        SelectedChatSession?.AddMessage("User", userInput);
        Debug.WriteLine($"User sent Message: {prompt}");
        var aiMessage = SelectedChatSession?.AddMessage("Assistant", string.Empty);

        IAsyncEnumerable<string> tokens = SelectedChatSession?.Executor.Inference(prompt, promptParameters?.BlackList ?? [],
            generationParameters ?? new TextGenerationParameters(true), _generation.CancelToken.Token);
        await foreach (var token in tokens!)
        {
            Debug.WriteLine($"RECEIVED TOKEN: {token}");

            if (token == "FILTERING MECHANISM TRIGGERED")
            {
                aiMessage!.Message = token;
                break;
            }

            aiMessage!.Message += token;
        }

        Debug.WriteLine($"Message: {aiMessage!.Message} Finished Receiving Tokens");

        _generation.CancelToken?.Dispose();
        _generation.State = GenerationState.Finished;
    }
```

**Comment by Pratham95VP** [comment]
try with` Dispatcher.Invoke` or` Dispatcher.InvokeAsync` . You can also try with ConfigureAwait(false).   `ChatService.Generate(userInput, GenerationParameters, PromptParameters).ConfigureAwait(false); `

**Reply by LSXAxeller** [reply]
non of them work, it still gets unresponsive the moment it call ChatService.Generate

**Comment by stevemonaco** [comment]
Some more investigative work needs to be done.

* Is it the `await foreach` loop that is causing the freeze?
* How many tokens are being processed?
* Are there a lot immediately available (I think `await foreach` here executes the next iteration immediately if available instead of via continuation)? 
* How big is `aiMessage.Message` getting? You probably want to use `StringBuilder` instead of continually building strings.
* Are there any UI updates from within `Generate` ? I think this is a no.
* `IsFirstRun(nameof(StatefulExecutorBase)).Result` if this is something async, then you want to `await`, but not enough context.

Either way, if your work is CPU-bound, you want to get that off of the UI thread.

**Reply by LSXAxeller** [reply]
in some weird case, it wasn't a problem with any of above, it was actually the usage of primary constructors in c#

**Reply by LSXAxeller** [reply]
my previous response was false report, I forget to get back `ChatService.Generate` instead `Task.Run(async () => await ChatService.Generate` in MessageSend.

I was using the same code in WPF with no freezing, I can't find what affects Avalonia


> * Is it the `await foreach` loop that is causing the freeze?

yes, more specifically await foreach (var token in tokens) where tokens `IAsyncEnumerable<string> tokens = _inference.InferAsync(prompt, messageParams, cancellationToken);` where _inference is ILLamaExecutor from LLamaSharp



> * How many tokens are being processed?

maybe 15 maybe 125 or even 1000, I can control the number of tokens I want to receive, but even 2 still giving same freeze behavior until it finish receiving tokens

> * Are there a lot immediately available (I think `await foreach` here executes the next iteration immediately if available instead of via continuation)?

usually it depends on the user hardware, my current hardware can generate 3 tokens/s, so yes it available


> * How big is `aiMessage.Message` getting? You probably want to use `StringBuilder` instead of continually building strings.

- > maybe 15 maybe 125 or even 1000, I can control the number of tokens I want to receive, but even 2 still giving same freeze behavior until it finish receiving tokens

token here is 1 word in most cases

> * Are there any UI updates from within `Generate` ? I think this is a no.

aiMessage itself inherented INotifyPropertyChanged so it's updating on every word, but the main problem that it even freeze before updating the UI on `ChatUserInput = string.Empty;` in MessageSend

> * `IsFirstRun(nameof(StatefulExecutorBase)).Result` if this is something async, then you want to `await`, but not enough context.

I did it, but it has no effect anyway or any impact, it just returns a Boolean

> Either way, if your work is CPU-bound, you want to get that off of the UI thread.

even using GPU backend for inferencing still getting the UI freeze

**Reply by stevemonaco** [reply, marked as answer]
This is a lot to sort through without the code. I'd suspect this is CPU-bound or at least I/O blocking at some level. I believe it goes through here at some point: https://github.com/SciSharp/LLamaSharp/blob/f1ec4aec34c7114397a7b13da50d3db4289449e1/LLama/LLamaContext.cs#L527

You can see the "Async" method just below is fake async. ie. It spins up a task which is not what an API is supposed to do. The caller is supposed to handle that in CPU-bound scenarios.

Besides that, the library isn't using best practices elsewhere. ie. https://github.com/SciSharp/LLamaSharp/blob/f1ec4aec34c7114397a7b13da50d3db4289449e1/LLama/LLamaStatelessExecutor.cs#L174 the `Infer` method itself has some `async` usage and yet it doesn't use `.ConfigureAwait(false)`. Neither is your code doing it as one of the services does: `await foreach (var token in modelSession.InferAsync(prompt, inferenceConfig, cancellationToken).ConfigureAwait(false))`

**Reply by LSXAxeller** [reply]
Thanks for investigating it, looks like like I need to go with the easy way "wrap in Task.Run" or fork the library in start digging into it
